### PR TITLE
Add train script that runs on the TPU VM

### PR DESCRIPTION
**Pull request by nostalgebraist**
Adds a script for training on the TPU VM, similar to the existing `device_sample.py`, etc.

My use case: I'm fine-tuning the model on a single TPU v3-8.  I initially got this working with the existing `train.py`, but

- I had to work around its expectations about file structure e.g. `meta.json`
- this way has lots of complexity that is needless here: using `ray`, using an extra VM outside the TPU, the `TPUCluster` wrapper, etc.

Note: I slapped this together hastily and it's not perfect.  I haven't yet verified it works on this branch, just on a personal dev branch with other changes.

### Comments

**Step**

- Compared to `train.py`, I changed when `step` is incremented, to try to make the values in the checkpoint file names match the exact number of gradient steps that have been run.
  - I increment the step after compiling model fn
  - Inside the loop, we do (update weights) --> (increment step) --> (save checkpoint if necessary), whereas train.py saves in between the update and the increment
- As a result of this, the val steps are now incremented by one, i.e. we'll do the first val at step 1 (because we took step 1 compiling model_fn)

**Data loader**

- The changes to the TFRecord reader class cover 2 issues:
  - On my small finetuning dataset, I'm doing multiple epochs.  After save/load, the iterator would recurse infinitely because every file was in the `used` list that was loaded.
    - Fixed with logic to discard `used` entries from earlier epochs
  - I added `reset()` method to try to make val data sampling deterministic.  (I have a tiny val dataset)

**Script**

- Does fine-tuning if `tune-model-path` is provided, otherwise trains from scratch (just for completeness)
- Compared to `train.py`, adds some more wandb metrics, print statements, etc.  You can take or leave these, they're just what I wanted for my project
- `save()` is mostly a copy-paste of `TPUCluster.save` which has some useful saving logic.  Ideally this logic would be defined only once, but I really wanted to avoid breaking `TPUCluster`

### CODE REVIEW

File: tfrecord_loader.py [hunk 1]
```diff
@@ -11,6 +11,10 @@ def __init__(self, index_fname, batch_size, parse_fn, map_fn=None, restore_state
             self.file_idx = restore_state["file_idx"]
             self.file_idx_init = False
             self.used = restore_state["used"]
+
+            unique = set(self.used)
+            last_pass_start_idx = len(self.used) % len(unique)
+            self.used = self.used[-last_pass_start_idx:]
```
Comment (kingoflolz) [inline tfrecord_loader.py:17]: I'm not sure why this is nessasary, can you reset self.used around the `except StopIteration:` line?

File: device_train.py [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+            print(f"No checkpoint to load at {initial_ckpt_path}. Training from scratch.")
+
+    if initial_ckpt_state_path:
+        print(f"path to load checkpoint from: {initial_ckpt_state_path}")
+    else:
+        print("not loading from a checkpoint")
+
+    # set up datasets
+    print("setting up datasets")
+    tpu_size = jax.device_count()
```
Comment (kingoflolz) [inline device_train.py:207]: you should probs check that tpu_size <= cores per replica here

File: device_train.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+
+    val_sets = {}
+
+    for k, v in params['val_set'].items():
+        val_sets[k] = TFRecordNewInputs(f"data/{v}",
+                                        batch_size=(global_val_batch,),
+                                        sample_size=seq)
+
+    # tok/sec metrics
+    windows_per_step = gradient_accumulation_steps * (per_replica_batch * tpu_size // cores_per_replica)
```
Comment (kingoflolz) [inline device_train.py:226]: I think sequences rather than windows is the more common terminology

**Review by kingoflolz — CHANGES_REQUESTED** [review]
Looks great, thanks for unwrapping a few layers of this onion and writing some simple code that trains properly haha. Would you be able to write up the config keys required either in the readme or with a commented fine tune config file?

Would also be good if you could write the rough tok/s you get on a v3-8 somewhere for reference

File: tfrecord_loader.py (hunk 1 above)
Comment (nostalgebraist) [reply tfrecord_loader.py]: Oh, yeah, that's a better idea

Changed it to just do `self.reset()` in that except block

File: device_train.py (hunk 2 above)
Comment (nostalgebraist) [reply device_train.py:204]: What situation would cause this check to fail?

(Trying to understand what I'm preventing so I can, e.g., write an error message about it)

Comment (kingoflolz) [reply device_train.py:204]: sometimes JAX doesn't detect the TPUs properly (and detects a single CPU device), or if there is a future model with more than 8 shards

Comment (nostalgebraist) [reply device_train.py:204]: Doesn't that mean we check that tpu_size >= cores per replica, not <=?

Comment (kingoflolz) [reply device_train.py:204]: Oh yes, my mistake

**Comment by kingoflolz** [comment]
Thanks!
